File: autoviz/examples/cpp/10_tutorial_log.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <chrono>
#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "common/tutorial_utils.hpp"

namespace {

constexpr int kDebug = 1;
constexpr int kInfo = 2;
constexpr int kWarn = 3;
constexpr int kError = 4;

struct Entry {
  int level;
  const char* name;
  const char* message;
};

const Entry kEntries[] = {
    {kDebug, "sensor", "imu sample ok"},
    {kInfo, "planner", "path length=12.4 m"},
    {kInfo, "controller", "tracking error=0.03 m"},
    {kWarn, "sensor", "lidar dropout 2 frames"},
    {kWarn, "planner", "goal near obstacle"},
    {kError, "controller", "cmd_vel timeout"},
    {kError, "autolink", "channel /cmd_vel no reader"},
};

std::string MakeLogJson(int level, const std::string& name,
                        const std::string& message, int line) {
  const auto now = std::chrono::system_clock::now().time_since_epoch();
  const auto ns =
      std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
  const int64_t sec = ns / 1000000000LL;
  const int64_t nsec = ns % 1000000000LL;
  return std::string("{\"timestamp\":{\"sec\":") + std::to_string(sec) +
         ",\"nsec\":" + std::to_string(nsec) + "},\"level\":" +
         std::to_string(level) + ",\"message\":\"" + message +
         "\",\"name\":\"" + name +
         "\",\"file\":\"10_tutorial_log.cpp\",\"line\":" +
         std::to_string(line) + "}";
}

}  // namespace

using autoviz::examples::MakeRawWriter;
using autoviz::examples::ParseRate;

int main(int argc, char** argv) {
  const double rate_hz = ParseRate(argc, argv, 2.0);

  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/log");
  auto writer = MakeRawWriter(node, "/fake/log", "foxglove.Log");

  autolink::Rate rate(rate_hz);
  std::cout << "log @ " << rate_hz << " Hz → /fake/log (foxglove.Log JSON)\n";

  constexpr int kN = static_cast<int>(sizeof(kEntries) / sizeof(kEntries[0]));
  int i = 0;
  while (autolink::OK()) {
    const Entry& e = kEntries[i % kN];
    const std::string msg =
        std::string(e.message) + " #" + std::to_string(i);
    const std::string json =
        MakeLogJson(e.level, e.name, msg, 40 + (i % 7));
    auto raw = std::make_shared<autolink::message::RawMessage>(json);
    writer->Write(raw);
    ++i;
    rate.Sleep();
  }
  return 0;
}

File: autoviz/examples/python/10_tutorial_log.py
#!/usr/bin/env python3
"""Emit foxglove.Log for Autoviz Log panel (level / name / search).

  /fake/log  foxglove.Log (JSON payload)

Log panel: Topic=/fake/log. Capture glog = Autoviz process only.

  /usr/bin/python3 examples/python/10_tutorial_log.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import argparse
import json
import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink

# foxglove.Log Level enum
DEBUG, INFO, WARN, ERROR = 1, 2, 3, 4

ENTRIES = (
    (DEBUG, 'sensor', 'imu sample ok'),
    (INFO, 'planner', 'path length=12.4 m'),
    (INFO, 'controller', 'tracking error=0.03 m'),
    (WARN, 'sensor', 'lidar dropout 2 frames'),
    (WARN, 'planner', 'goal near obstacle'),
    (ERROR, 'controller', 'cmd_vel timeout'),
    (ERROR, 'autolink', 'channel /cmd_vel no reader'),
)


def make_log(level: int, name: str, message: str, line: int) -> bytes:
    now = time.time_ns()
    return json.dumps(
        {
            'timestamp': {'sec': now // 1_000_000_000, 'nsec': now % 1_000_000_000},
            'level': level,
            'message': message,
            'name': name,
            'file': '10_tutorial_log.py',
            'line': line,
        },
        separators=(',', ':'),
    ).encode()


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--rate', type=float, default=2.0)
    args = ap.parse_args()

    autolink.init('log')
    n = autolink.Node('/autoviz/log')
    w = n.create_writer('/fake/log', 'foxglove.Log', qos_depth=20)
    rate = autolink.Rate(args.rate)
    print(f'log @ {args.rate} Hz → /fake/log (foxglove.Log JSON)')

    i = 0
    try:
        while not autolink.is_shutdown():
            level, name, msg = ENTRIES[i % len(ENTRIES)]
            w.write(make_log(level, name, f'{msg} #{i}', 40 + (i % 7)))
            i += 1
            rate.sleep()
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()
